### ISSUE TITLE
Questions before contributing more to the project

### ISSUE THREAD

**Issue by jentfoo** [sorcix/sIRC#16]
I definetly like your guys library a lot.  I have gone through the code, and love the simplcity of it.  I run an IRC network, and would be curious about using it as a foundation for my IRC network, but I would like to continue to improve things before I start relying on it.

So I was curious about your perspective on a few things.

1) What is your policy about API changes?  If I were to recommend some API changes how easy or interested would you be in seeing them?

2) Right now IrcOutput and IrcInput are blocked threads...while this is not a big deal if the library is used for only a few connections, it would be nicer to use a thread pool and have things done via asynchronis callbacks.  At least that's my opinion, and opinions definetly may vary, so I am just curious what your guys thoughts are on this?  The reason I asked the first question is this would likely require API changes to pass in thread pool (though that can be worked around if desired).

3) What's your policy or opinion on using additional external libraries?  I can understand wanting to keep dependencies low, but I see room for improvement.  

For example in when iterating over the "ServerListeners" to call on to actions, if a listener throws an exception it may prevent other listeners from being called.  Or if a listener either removes or adds a listener from the result of being called (for example maybe a listener only wants to be called for one event, then removes itself after), a ConcurrentModificationException will be thrown as it iterates to other listeners.

This is just a specific example of a problem that is solved in another library and you could leverage to your advantage, if I apply some of the threading changes, there will be others as well.  So I just want to get a feel of your opinion on adding additional dependenices.

Thank you guys for your feedback, I would love to help improve your library!

**Comment by JamoBox** [comment]
I'm only a contributor on this project so don't go by what I say, but here's my 2 cents on your questions;

1) As with all API changes they should be made backwards compatible where possible if changing existing public facing calls. As for additions to the API, so long as it is well documented and does not cause any issues then I can see no problem there.

2) It's been a while since I've looked at these and it is apparent that there is a lot of optimization that can be done on them. I guess scalability is the issue here, and it would be nice to see some improvements on that end so again, I'm sure this would be welcomed (given that answer to question 1 is taken into account).

3) My only issues with adding dependencies is potentially limiting what the library can run be used with, and how much space sIRC takes up as a library itself. However as the first issue usually only really applicable to badly written or non-mainstream libraries, I can't see that mattering too much. That only leaves the second issue; generally speaking, so long as the benefits of the library justify it's  inclusion in sIRC then it should not be a problem.

It's nice to see some more interest in this project, I look forward to seeing what @sorcix has to say concerning this.

**Comment by sorcix** [comment]
Hello there!

1) I agree with @JamoBox here, public facing API should be backwards compatible unless there is a very good reason to change it. Additions are fine, as long as they stay within scope of the project being an IRC library. I have made some mistakes there myself, things like the SIRCService class shouldn't have been in here.

2) My use case was a single IRC client on a single network, so you're right when you say that there may be better options when opening a lot of connections. I'm open for this change, as long as it doesn't have too much negative effects/overhead on projects that only use a single connection. The IrcInput and IrcOutput classes are not part of the public API, so backwards compatibility is less important here.

3) I think simplicity is one of the main selling points for sIRC, and not having too much external dependencies is part of that. That said, I'm open for using external libraries as long as it's justified. As long as we don't include a giant dependency for a single method, it's fine for me.

I hope to get some comments from @pfn as well, as I know he's using sIRC for an Android client and may be affected by possible changes.

**Comment by jentfoo** [comment]
Thanks for the great feedback.  Maybe it will help if I describe some of my motivations.

1) I asked about API changes because if I am making threading changes, it is obviously ideal to be able to provide your own thread pool.  Now that I think about it this can be easily done with just expanding the API (allowing you to pass in an Executor if desired, and if one is not passed in we could construct and use a default one).  So I don't think this is going to particularly be an issue, the API changes that likely wont be backwards compatible would be around IrcInput, IrcOutput, IrcQueue (as the queue can no longer block), etc.

2) That's pretty cool, I honestly never even considered using this as a foundation for a client.  I personally have been looking at this as building and running a series of bots on my IRC network.  I don't see any reason that this library can't offer both.

3) Specifically the library I was thinking of introducing is one me and a few colleges work on called threadly: https://github.com/threadly/threadly

It's available via maven central, and is really quite small (just over 200KB).  It does not have any additional dependencies on top of it.  I pointed out a small example of where it could help simplify and fix things in the existing implementation.  But as I build out a slightly different threading design, I think it will be very useful there as well.

I definitely don't want to introduce it if it bothers you guys.  I 100% understand wanting to keep a library small and with minimal dependencies (which is why threadly has no external dependencies).  I just wanted to ask and get your take on it.

Thanks for all the great feedback guys!

**Comment by pfn** [comment]
I also prefer few API changes, but there are some fundamental things that need to be changed that might end up necessitating public API changes anyway.
1. IRC capability negotiation
   a. this includes stuff like multiprefix (for multi-status on users)
   b. SASL authentication
   c. server-time
2. custom parsing, this is needed to re-enable color codes
3. a more varied/powerful callback api for IRC events (e.g. server-time from previous, also to timestamp individual packets)

All these features require a fair amount of public facing change. It might be possible to keep some of it backward compatible, but it could be difficult.

**Comment by JamoBox** [comment]
It will probably be worth creating a develop branch from master, and use that as a base branch for the API changes so there is a clear overview of what has changed that can be tested before merging into master.

**Comment by jentfoo** [comment]
Of course.  I will create a branch and propose something via a pull request.  No obligations of course, but at least you can take it if you like, or leave it if you don't.

Thanks for all the great feedback everyone.

**Comment by JamoBox** [comment]
I meant as in a branch on this main project that can be used for changes concerning this ticket. That way with multiple people working on these changes they don't merge into master until they have been collectively tested. This would require @sorcix to create the develop branch on this repo.

**Comment by sorcix** [comment]
I have created the develop branch.
